ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceCall.java

```diff
@@ -0,0 +1,251 @@
+    return new TraceOptions(
+        traceTypes.contains(TraceTypeParameter.TraceType.STATE_DIFF),
+        false,
+        traceTypes.contains(TraceTypeParameter.TraceType.TRACE)
+            || traceTypes.contains(TraceTypeParameter.TraceType.VM_TRACE));
+  }
+
+  private void parseRevertReasonToOutput(
+      final Bytes revertReason, final ObjectNode resultNode, final ObjectMapper mapper) {
+    if (!resultNode.findValue(TRACE_NODE_PROPERTY_NAME).isEmpty()) {
+      final ArrayNode traceNode = mapper.createArrayNode();
+      ObjectNode json = mapper.valueToTree(resultNode.findValue(TRACE_NODE_PROPERTY_NAME).get(0));
+
+      final JsonNode revertReasonNode = json.findValue("revertReason");
+      if (revertReasonNode != null && !revertReasonNode.textValue().isBlank()) {
+        json.remove("revertReason");
+        traceNode.add(json);
+        resultNode.replace(TRACE_NODE_PROPERTY_NAME, traceNode);
+      }
+    }
+
+    final TextNode outputWithRevertReason =
+        mapper.createArrayNode().textNode(revertReason.toHexString());
+    if (resultNode.findValue(OUTPUT_NODE_PROPERTY_NAME).textValue().equals("0x")) {
+      resultNode.replace(OUTPUT_NODE_PROPERTY_NAME, outputWithRevertReason);
+    }
+  }
+
+  private Optional<BlockHeader> resolveBlockHeader(
+      final Optional<BlockParameter> maybeBlockParameter) {
+    AtomicLong blockNumber = new AtomicLong();
+
+    maybeBlockParameter.ifPresentOrElse(
+        blockParameter -> {
+          if (blockParameter.isNumeric()) {
+            blockNumber.set(blockParameter.getNumber().get());
+          } else if (blockParameter.isEarliest()) {
+            blockNumber.set(0);
+          } else if (blockParameter.isPending() || blockParameter.isLatest()) {
+            blockNumber.set(blockchainQueries.get().headBlockNumber());
+          } else {
+            throw new IllegalStateException("Unknown block parameter type.");
+          }
+        },
+        () -> blockNumber.set(blockchainQueries.get().headBlockNumber()));
+
+    return blockchainQueries.get().getBlockHeaderByNumber(blockNumber.get());
+  }
+
+  private void generateTracesFromTransactionTrace(
+      final TraceWriter writer,
+      final ProtocolSchedule protocolSchedule,
+      final TransactionTrace transactionTrace,
+      final Block block,
+      final TraceFormatter formatter,
+      final AtomicInteger traceCounter) {
+    formatter
+        .format(protocolSchedule, transactionTrace, block, traceCounter)
+        .forEachOrdered(writer::write);
+  }
+
+  private void setNullNodesIfNotPresent(final ObjectNode parentNode, final String... keys) {
+    Arrays.asList(keys)
+        .forEach(
+            key ->
+                Optional.ofNullable(parentNode.get(key))
+                    .ifPresentOrElse(ignored -> {}, () -> parentNode.put(key, (String) null)));
+  }
+
+  private void setEmptyArrayIfNotPresent(final ObjectNode parentNode, final String... keys) {
```

Comment:
same as above

Reply:
Good advice. I've added a result class and it made the reduced the complexity quite a lot.